regex-syntax/src/hir/interval.rs

```diff
@@ -347,24 +347,19 @@ impl<I: Interval> IntervalSet<I> {
         self.ranges.sort();
         assert!(!self.ranges.is_empty());
 
-        // Is there a way to do this in-place with constant memory? I couldn't
-        // figure out a way to do it. So just append the canonicalization to
-        // the end of this range, and then drain it before we're done.
-        let drain_end = self.ranges.len();
-        for oldi in 0..drain_end {
-            // If we've added at least one new range, then check if we can
-            // merge this range in the previously added range.
-            if self.ranges.len() > drain_end {
-                let (last, rest) = self.ranges.split_last_mut().unwrap();
-                if let Some(union) = last.union(&rest[oldi]) {
-                    *last = union;
-                    continue;
-                }
+        // We consistently try to merge range with previous range
+        // and merge them if possible. Otherwise, we make it the
+        // range as the last one.
```

Comment:
Could you clean this comment up please? It's a little hard to parse. A high level comment explaining why this works would also be helpful.

Reply:
Just rewrote the comments, hope it is more clear.